### ISSUE TITLE
provide fully flow-based grunt-tasks  (was: directory_tree: use per-artifact configuration)

### ISSUE THREAD

**Issue by x1B** [LaxarJS/grunt-laxar#41]
### Status Quo (v1.0.0)

Currently, the _Gruntfile_ configuration for your average LaxarJS project contains a lot of paths and patterns that are specific to the artifacts used by the application and to their relative location. Especially if artifacts are installed as *bower_components*, things can get tricky. This may lead to subtle, hard-to-diagnose errors where resources cannot be found, or where their correct theme folder is not used.

Also, the `watch` task for live reload may become very slow, just by having the wrong set of files somewhere within your project (e.g. a nested *node_modules* somewhere deep in a widget), due to missing exclusion rules.

Ideally, we would be able to have a Gruntfile that is simplified to the point where there are (almost) only _zero-configuration defaults_  that _just work_, while making sure that everything, especially live reload, is handled in a responsive manner. 

### The Ansatz

Based on which artifacts are reachable from a project's *flow* we should determine what files to *watch*, what files to _include/embed_ into the listings, and what JS *modules* to load as dependencies.

For this, we add a new task `laxar-artifacts` which creates a model of your entire application based on a set of flows to include. Starting at the flows, the task identifies *pages*, *compositions* and *layouts* and their resources, taking into account the set of installed *themes*. Next, all _widgets_ on these pages and their _controls_ are identified. The resulting model is stored in JSON format, and should be useful not just for other grunt-tasks, but possibly for client-side tooling as well.

To allow for additional resources to be watched/listed/included, the artifact's _descriptor_ (widget.json, control.json) will support a new top-level entry `resources` with the following schema:

```js
{
   type: "object",
   properties: {
      "watch": { 
         type: "array", 
         items: { type: "string" }
      },
      "list": { 
         type: "array", 
         items: { type: "string" }
      },
      "embed": { 
         type: "array", 
         items: { type: "string" }
      }
   }
}
```

Usually, it is *not necessary* to add this information per-artifact, as the defaults will almost always be sufficient.

### New Flow-Based Artifacts

Now that we have a complete model of our flow artifacts, we can start using it to generate the following things:

* _dependencies_ used by RequireJS (and r.js for production) to find and load AMD modules that implement widgets and controls

* _resources_ mainly used by the LaxarJS runtime to determine for which artifact HTML/CSS resources are available, and in which theme

* _watcher configuration_ used by the development server to know when to reload the application in the browser

* _merged CSS and JavaScript_ for all artifacts used in a flow, so that CSS is loaded quickly in production.

Because each set of flows is handled in isolation, it is possible to split your application into several flows which select from the same set of artifacts. For example, there might be one flow for the regular use of your application, and another flow for configuring and customizing the application. 

### Flow-Aware Tasks for grunt-laxar v1.1.0

To implement all of this, a new set of flow-aware tasks is needed. Task output is stored under `var/flows/$flowSet` where `$flowSet` is a configurable identifier that corresponds to a set of flow definitions. By default (using the application template), there is one flow-set `main` containing the flow `application/flows/flow.json`. Here are the individual tasks with their outputs:

- [x] `laxar-artifacts` collects all artifacts from the set of flows, creating an application model for them (`tooling/artifacts.json`)

- [x] `laxar-dependencies` collects AMD modules that implement artifacts (`dependencies.js`)

- [x] `laxar-resources` generates the resource listings for these artifacts based on which of the associated _list_ and _embed_ resources are _actually available_ on the file system (`resources.json`)

- [x] `laxar-test` runs karma/PhantomJS and _jshint_ on all widgets used by the flow (`tooling/test-results.xml`).

- [x] `laxar-watch` configures and launches the grunt `watch` task based on the set of flows and stores the configuration for inspection (`work/watch-configuration.json`, *development only*)

- [x] `laxar-dist-css` creates a combined CSS file for this flow, for each theme (`dist/my.theme.css`, *production only*)

- [x] `laxar-dist-js` uses *r.js* to produce a single JavaScript file for the flow and all its dependencies (`dist/bundle.js`/`dist/bundle.js.map`, *production only*).

- [x] `laxar-configure` takes a set of flows and uses it to configure all the tasks above, once for production, and once for development (`work/task-configuration.json`).

### Compatibility with v1.0.0

Although it should be very simple to upgrade any project to using the new tasks, the old tasks will be kept for compatibility with LaxarJS v1.0.0, but may issue deprecation warnings in the future.

**Comment by x1B** [comment]
### Updated the description

- Added lots of information on why and how the task system will change

- The previous description mentioned a _resources.json_ for listing the files. Instead, we changed the requirements to instead allow for the _widget.json_ and _control.json_ to contain such information, to avoid unnecessary clutter. In the future, support for a _layout.json_ and a _theme.json_ might be added, if there are tangible benefits.

- Added checkboxes to track task progress

**Comment by jpommerening** [comment]
Sounds great! I like the resource- and action-centric names of the tasks a lot better than what we had before.

Just a heads up: Maybe we can use an existing solution (other than our home-grown `css_merger`) for merging CSS to gain Source-Map support. [clean-css](https://www.npmjs.com/package/clean-css) looks like it's up to the job

**Comment by x1B** [comment]
Thanks, _clean-css_ looks very promising as it already seems to support URL rebasing of `url(...)` property values. I'll make sure to check it out.

**Comment by x1B** [comment]
### Updated the Description

Added `laxar-test` task. Also, the following boxes will need to be checked before this feature can be merged:

- [ ] provide comprehensive _documentation_ on all tasks

- [ ] make sure all tasks are well-covered by _tests_

**Comment by x1B** [comment]
So, things are already going pretty smooth with the new setup, but there are two things we forgot:

- [ ] currently, there is usually a single `init` entry file. We need a naming scheme to have multiple files (for multiple flows). Or maybe we should simply generate these files? In this case we would need a way to specify additional modules in the flow-target-configuration.

- [ ] in production, the LaxarJS runtime tries to load CSS from `var/static/css`, unless a `link` element with `data-ax-merged-css` lives on the page. Having that element is highly recommended anyway, for performance reasons. So maybe we just have to make this mandatory for using the new tasks.

- [ ] the `test_results_merger` functionality is not yet provided by `laxar-test`.

**Comment by x1B** [comment]
Regarding the `init` problem: The following project-artifacts from a default template-based project currently make assumptions on the flow, by directly referencing it, or indirectly by depending on flow-specific artifacts:

- `require_configuration.js` (direct reference)
- `debug.html` (reference to `require_configuration.js`, and to `init.js`)
- `index.html` (reference to the flow-specific `bundle.js` and merged `x.theme.css`).
- `init.js` (reference to flow-specific `dependencies.js` and `resources.json`)

All of these files have application-specific components to them as well...

**Comment by x1B** [comment]
Implemented on `master` (`v1.1.0`).

### Upgrade Information

Most of the existing grunt-laxar tasks have been _deprecated_ in favor of a new set of _flow-based_ grunt tasks. The new tasks have the following advantages:

- much less configuration
- better performance due to more specific watchers
- basic support for multiple flows within one project
- no restart required after adding a new widget
- using _laxar-_ naming prefix for compatibility with other tasks
- do not depend on the `laxar` package

The old tasks are still available, but upgrading is relatively easy (no pages or widgets need to be modified). So upgrading is strongly recommended starting with grunt-laxar v1.1.0.


#### 1. Gruntfile

To upgrade your Gruntfile, **remove all grunt-laxar related configuration** and task aliases. Then add configuration for the task _laxar-configure:_

```js
      'laxar-configure': {
         options: {
            flows: [
               { target: 'main', src: 'application/flow/flow.json' }
            ],
            ports: {
               develop: 8000,
               test: 9000,
               livereload: 35729
            }
         }
      }
```

Modify the ports as needed. You can omit the `ports` configuration (ports above are the default values). Also, you probably want to add the following alias tasks:

```js
   grunt.loadNpmTasks( 'grunt-laxar' );

   // basic aliases
   grunt.registerTask( 'test', [ 'laxar-test' ] );
   grunt.registerTask( 'build', [ 'laxar-build' ] );
   grunt.registerTask( 'dist', [ 'laxar-dist' ] );
   grunt.registerTask( 'develop', [ 'laxar-develop' ] );

   // additional (possibly) more intuitive aliases
   grunt.registerTask( 'optimize', [ 'laxar-dist' ] );
   grunt.registerTask( 'start', [ 'laxar-develop' ] );
   grunt.registerTask( 'start-no-watch', [ 'laxar-develop-no-watch' ] );
   grunt.registerTask( 'default', [ 'build', 'test' ] );
```

#### 2. The _init_ module

Your project should have a JavaScript _entry module_, usually called _init.js_.
Because _grunt-laxar_ uses a different directory structure, you need to adjust your AMD dependencies. Here is an example:

```js
require( [
   'laxar',
   'laxar-application/var/flows/main/dependencies',
   'json!laxar-application/var/flows/main/resources.json'
], function( ax, mainDependencies, mainResources ) {
   'use strict';
   window.laxar.fileListings = { application: mainResources, includes: mainResources };
   ax.bootstrap( mainDependencies );
} );
```


#### 3. The _index.html_ entry

For development, you will usually use the _debug.html_ that is part of the LaxarJS application template. Because this file uses optimized assets, paths need to be adjusted as well:

```
<head>
   <!-- ... -->
   <!-- always use a link element in production, and point it to the new path -->
   <!-- if using a custom theme, substitute it for 'default.theme' -->
   <link data-ax-merged-css href="var/flows/main/dist/default.theme.css" rel="stylesheet" type="text/css">
</head>
<body>
   <!-- ... -->
   <div data-ax-page></div>
   <div data-ng-view></div>

   <script data-ax-application-mode="RELEASE" src="application/application.js"></script>
   <!-- this path has changed as well -->
   <script src="var/flows/main/dist/bundle.js" async></script>
</body>
```

### More Information

For more information, consult the updated [grunt-laxar documentation](https://github.com/x1B/grunt-laxar/blob/master/README.md#grunt-laxar).
